### PR TITLE
Memoryless storage mode with fallback (Metal)

### PR DESCRIPTION
**Pull request by veghbalazs**
This PR changes the storage mode of MSAA sidecar textures to [memoryless](https://developer.apple.com/documentation/metal/mtlstoragemode/memoryless) in Metal backend, on supported devices. 

Memoryless sidecars had been attempted in the past, but reverted in https://github.com/google/filament/pull/4333 due to a sparsely documented limitation of this feature: only a certain geometry complexity can be processed and rendered this way. Exceeding this limit causes a command buffer error.

We conducted a few tests and concluded that we're unable to predict this abstract "memoryless limit". It does not only differ between Apple GPU generations (e.g. A12X or M1 Pro), but it's changing based on the resolution of the rendertarget and available system memory.

Therefore, we decided to resort to a simple fallback mechanism. It handles the first memoryless limit reached event by turning off the feature and falling back to private storage mode. The user will experience 2-3 dropped frames, but their app will continue normal operation from then on (with increased memory usage understandably). Note that no attempts are made to return to memoryless, an engine restart is needed.

### CODE REVIEW

File: filament/backend/src/metal/MetalHandles.mm [hunk 1]
```diff
@@ -1135,7 +1135,7 @@ BufferDescriptor b(nullptr, 0u, [](void* buffer, size_t size, void* user) {
     return MTLStoreActionStore;
 }
 
-id<MTLTexture> MetalRenderTarget::createMultisampledTexture(id<MTLDevice> device,
+id<MTLTexture> MetalRenderTarget::createMultisampledTexture(MetalContext& context,
```
Comment (bejado) [inline filament/backend/src/metal/MetalHandles.mm:1138]: Instead of passing `MetalContext&` to this method, let's simply make it non-static.

**Review by bejado — CHANGES_REQUESTED** [review]
Thanks! Overall LGTM, just a few things.

File: filament/backend/src/metal/MetalContext.h [hunk 2]
```diff
@@ -58,8 +59,11 @@ struct MetalContext {
     id<MTLCommandBuffer> pendingCommandBuffer = nullptr;
     id<MTLRenderCommandEncoder> currentRenderPassEncoder = nullptr;
 
+    std::atomic<bool> memorylessLimitsReached = false;
```
[the remark this answers is not in the document]
Comment (veghbalazs) [reply filament/backend/src/metal/MetalContext.h:63]: fixed in https://github.com/google/filament/pull/6124/commits/f9354500cfd72223f491a7d1abe9e19ac6d077cf

File: filament/backend/src/metal/MetalDriver.mm [hunk 3]
```diff
@@ -92,6 +92,9 @@
             mContext->highestSupportedGpuFamily.mac   >= 2;     // newer macOS GPUs
     }
 
+    // In order to support memoryless render targets, an Apple GPU is needed.
+    mContext->supportsMemorylessRenderTargets = mContext->highestSupportedGpuFamily.apple >= 1;
```
[the remark this answers is not in the document]
Comment (veghbalazs) [reply filament/backend/src/metal/MetalDriver.mm:98]: added a comment in https://github.com/google/filament/pull/6124/commits/f9354500cfd72223f491a7d1abe9e19ac6d077cf

File: filament/backend/src/metal/MetalHandles.mm (hunk 1 above)
Comment (veghbalazs) [reply filament/backend/src/metal/MetalHandles.mm]: 👍  changed to non-static in https://github.com/google/filament/pull/6124/commits/f9354500cfd72223f491a7d1abe9e19ac6d077cf

**Review by bejado — APPROVED** [review]

**Pull request merged**